ocaml/testsuite/tests/typing-layouts-or-null/variables.ml

```diff
@@ -397,3 +397,17 @@ Error: This type t_value_or_null dummy should be an instance of type 'a dummy
        But the kind of t_value_or_null must be a subkind of value
          because of the definition of constrained' at lines 1-2, characters 0-44.
 |}]
+
+(* Copied from the tree, should work. *)
+let should_work s =
```

Comment:
I'm surprised this didn't work out of the box. It's true that the use of `a` and `b` in the body of the object constrain their kinds, but since `a` and `b` are inferred to have completely unconstrained types, this should be fine.

I don't think changing the capture requirement for objects is the right fix here. (Though perhaps it is expedient to do so to get nightly unbroken.)

Reply:
Type variables for `a` and `b` are generalized before the check happens. Perhaps we should rework the check to happen before, but I think for now we should just allow `maybe_null` to fix tree builds.

Reply:
That still should be OK. We get `a : ('a : value_non_null). 'a` and `b : ('b : value_non_null). 'b`. But then the occurrences of `a` and `b` should get instantiated to be values: instantiations can always be more specific.

Or maybe you're saying that the method variables `a` and `b` get generalized too soon? That could indeed cause a problem.... but I would expect methods to always need to be `value`s, too. So I still think everything should be OK here.

Reply:
No, we get `a : ('a : value) . 'a` when we unpack the tuple, since tuple elements can be null. And then `'a` is generalized.

Reply:
Oops. I mistyped in my comment. I meant that the type variables would be `value_or_null` (or is that `value_maybe_null` these days?). The rest of the comment stands.

Reply:
Resolved per conversation.